text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+Finally, there are still a few features lacking in the compiler that will add additional hurdles to implementation, like:
+
+* a lack of const-generic bounds, like `N >= 8`
+* the lack of [generic const expressions][`generic_const_exprs`], like `[u8; {N.div_ceil(8)}]`
+
+However, this is substantially fewer hurdles than last time, and more cases have been brought up where generic integers will be useful despite these.
+
+[pattern types]: https://github.com/rust-lang/rust/issues/123646
+[#2581]: https://github.com/rust-lang/rfcs/pull/2581
+[Zig]: https://ziglang.org/documentation/master/#Primitive-Types
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Use `uN` and `iN` instead of `u<N>` and `i<N>`
+
+Since one of the major benefits of the generic types is to be able to use them in generic code, this isn't an option.
+
+However, this RFC doesn't actually stop `uN` and `iN` from being added as aliases to `u<N>` and `i<N>` in the future. While the version with an explicit generic parameter is required for generic code, it should in theory be possible to add these aliases if people want them. These would require special changes to the language itself, since it does not make sense to add `2.pow(32)` items to the prelude.
+
+## Use `uint<N>` and `int<N>` instead of `u<N>` and `i<N>`
+
+This was the previous proposal used by the original RFC, and they were chosen due to a misconception about name conflicts. Although the `i` variable is commonly used for loops and other index variables, variable names and type names are entirely separate and can coexist peacefully. Having names closer to the originals feels more discoverable here.
+
+Additionally, the names `uint` and `int` remain available for future, potentially unrelated integer types.
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way to generalise integers *for computers*, but generalising based upon bounds is a very natural way for humans to do it, *and* more general. For example, instead of `u<N>` and `i<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a percentage could be represented exactly as `int<0..=100>`. Whether an integer is signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot different from the existing integer types in the language. Additionally, such a proposal can coexist with this one, since there's nothing stopping us from making an `int<A..=B>` type in the future such that `u<N>` and `i<N>` are equivalent to the correct range of integers.
+
+## Integer traits
```

Comment:
I don't think it would be a good alternative for the standard library to provide such traits. However, they do exist in the crates.io ecosystem and are reasonably popular (mainly num-traits). Insofar the RFC is motivated by letting people generalize *their* code over the existing primitive types, it would benefit from engaging with those traits. There's definitely things a language feature can do better, e.g., `impl<N> Foo for u<N>` is less likely to cause coherence problems than `impl<T: Int> Foo for T`. On the other hand, the existing traits do solve some problems, and not all drawbacks listed in this section apply to them. They can iterate on "which methods are in which trait" and "which set of traits even exists" much more easily because semver-breaking changes are possible. It's also possible to have competing traits with overlapping sets of methods, which is mitigated by std providing inherent impls (though there can still be ambiguities: https://github.com/rust-lang/rust/issues/68826).